packages/connect/src/connect-error.ts

```diff
@@ -95,6 +95,72 @@ export class ConnectError extends Error {
     this.details = outgoingDetails ?? [];
     this.cause = cause;
   }
+
+  /**
+   * Convert any value - typically a caught error into a ConnectError,
+   * following these rules:
+   * - If the value is already a ConnectError, return it as is.
+   * - If the value is an AbortError from the fetch API, return the message
+   *   of the AbortError with code Canceled.
+   * - For other Errors, return the error message with code Unknown by default.
+   * - For other values, return the values String representation as a message,
+   *   with the code Unknown by default.
+   */
+  static from(reason: unknown, code = Code.Unknown): ConnectError {
+    if (reason instanceof ConnectError) {
+      return reason;
+    }
+    if (reason instanceof Error) {
+      if (reason.name == "AbortError") {
+        // Fetch requests can only be canceled with an AbortController.
+        // We detect that condition by looking at the name of the raised
+        // error object, and translate to the appropriate status code.
+        return new ConnectError(reason.message, Code.Canceled);
+      }
+      return new ConnectError(reason.message, code);
+    }
+    return new ConnectError(String(reason), code);
+  }
+
+  /**
+   * Retrieve error details from a ConnectError. On the wire, error details are
+   * wrapped with google.protobuf.Any, so that a server or middleware can attach
+   * arbitrary data to an error. This function decodes the array of error details
+   * from the ConnectError object, and returns an array with the decoded
+   * messages. Any decoding errors are ignored, and the detail will simply be
+   * omitted from the list.
+   */
+  findDetails<T extends Message<T>>(type: MessageType<T>): T[];
+  findDetails(registry: IMessageTypeRegistry): AnyMessage[];
+  findDetails(
+    typeOrRegistry: MessageType | IMessageTypeRegistry
+  ): AnyMessage[] {
+    const registry =
+      "typeName" in typeOrRegistry
```

Comment:
I'm surprised we don't have an `IsMessage` type guard yet.  we have to copy `"typeName" in` three times because of it.  It'd make this code much clearer (and less error prone since there's no exclusivity check on the `in` operator which means it'll let you check things that don't exist in either type, e.g. `typename`).  I'll put up a protobuf-es PR to add `IsMessage`.

Reply:
Hold on. We wouldn't want to use a generic type guard function here because it would come with a much larger bundle size increase. How often do you need to distinguish a message _type_? I do not believe there is a pressing need for a guard.